File: java/dsa-java-main1/try/heap.java
import java.util.ArrayList;

class MaxHeap {
    // Function to heapify a subtree rooted at index i
    void heapify(ArrayList<Integer> heap, int size, int i) {
        int largest = i;
        int left = 2 * i + 1;
        int right = 2 * i + 2;

        // Check if left child is larger than root
        if (left < size && heap.get(left) > heap.get(largest))
            largest = left;

        // Check if right child is larger than root
        if (right < size && heap.get(right) > heap.get(largest))
            largest = right;

        // If the largest is not the root, swap and recursively heapify
        if (largest != i) {
            // Swap elements at i and largest
            int temp = heap.get(largest);
            heap.set(largest, heap.get(i));
            heap.set(i, temp);

            // Recursively heapify the affected sub-tree
            heapify(heap, size, largest);
        }
    }

    // Function to insert a new element into the max-heap
    void insert(ArrayList<Integer> heap, int newNum) {
        int size = heap.size();
        heap.add(newNum);

        // Heapify from the bottom to the top
        for (int i = size / 2 - 1; i >= 0; i--) {
            heapify(heap, size, i);
        }
    }

    // Function to delete a specific element from the max-heap
    void deleteNode(ArrayList<Integer> heap, int num) {
        int size = heap.size();
        int i;

        // Find the index of the element to be deleted
        for (i = 0; i < size; i++) {
            if (num == heap.get(i))
                break;
        }

        // Swap the found element with the last element
        int temp = heap.get(i);
        heap.set(i, heap.get(size - 1));
        heap.set(size - 1, temp);

        // Remove the last element
        heap.remove(size - 1);

        // Heapify the remaining elements
        for (int j = size / 2 - 1; j >= 0; j--) {
            heapify(heap, size - 1, j);
        }
    }

    // Function to print the elements of the array
    void printArray(ArrayList<Integer> array, int size) {
        for (Integer i : array) {
            System.out.print(i + " ");
        }
        System.out.println();
    }

    public static void main(String args[]) {
        ArrayList<Integer> array = new ArrayList<Integer>();
        int size = array.size();

        MaxHeap maxHeap = new MaxHeap();
        maxHeap.insert(array, 3);
        maxHeap.insert(array, 4);
        maxHeap.insert(array, 9);
        maxHeap.insert(array, 5);
        maxHeap.insert(array, 2);

        System.out.println("Max-Heap array: ");
        maxHeap.printArray(array, size);

        maxHeap.deleteNode(array, 4);
        System.out.println("After deleting an element: ");
        maxHeap.printArray(array, size);
    }
}

File: java/dsa-java-main1/try/heap.py
# Function to heapify a subtree rooted at node i
def heapify(arr, size, root):
    largest = root
    left = 2 * root + 1
    right = 2 * root + 2

    if left < size and arr[root] < arr[left]:
        largest = left

    if right < size and arr[largest] < arr[right]:
        largest = right

    if largest != root:
        arr[root], arr[largest] = arr[largest], arr[root]
        heapify(arr, size, largest)

# Function to insert a new element into the max-heap
def insert_into_max_heap(heap, new_element):
    size = len(heap)

    if size == 0:
        heap.append(new_element)
    else:
        heap.append(new_element)
        for i in range((size // 2) - 1, -1, -1):
            heapify(heap, size, i)

# Function to delete a specific element from the max-heap
def delete_from_max_heap(heap, element_to_delete):
    size = len(heap)
    i = 0

    for i in range(0, size):
        if element_to_delete == heap[i]:
            break

    heap[i], heap[size - 1] = heap[size - 1], heap[i]
    heap.pop()

    for i in range((len(heap) // 2) - 1, -1, -1):
        heapify(heap, len(heap), i)

# Example usage
max_heap = []
insert_into_max_heap(max_heap, 3)
insert_into_max_heap(max_heap, 4)
insert_into_max_heap(max_heap, 9)
insert_into_max_heap(max_heap, 5)
insert_into_max_heap(max_heap, 2)

print("Max-Heap array: " + str(max_heap))

delete_from_max_heap(max_heap, 4)
print("After deleting an element: " + str(max_heap))
